### PR TITLE
APP-907 / APP-908 / APP-909 / APP-910

### PR DESCRIPTION
**Pull request by robson-sanchez**
- APP-907 Define a templated PresentationML for issue created event
- APP-908 Implement an engine to build the EntityJSON according to the metadata
- APP-909 Define metadata objects to build an EntityJSON for issue created event
- APP-910 Implements JIRA Parser to build the message to be posted

### CODE REVIEW

File: src/main/java/org/symphonyoss/integration/webhook/jira/parser/V1ParserFactory.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parser factory for the MessageML v1.
+ * Created by rsanchez on 21/03/17.
+ */
+@Component
+public class V1ParserFactory extends JiraParserFactory {
```
Comment (crepache) [inline src/main/java/org/symphonyoss/integration/webhook/jira/parser/V1ParserFactory.java]: Is it possible to move to /v1 and rename ParserFactory ?

File: src/main/java/org/symphonyoss/integration/webhook/jira/parser/V2ParserFactory.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parser factory for the MessageML v2.
+ * Created by rsanchez on 21/03/17.
+ */
+@Component
+public class V2ParserFactory extends JiraParserFactory {
```
Comment (crepache) [inline src/main/java/org/symphonyoss/integration/webhook/jira/parser/V2ParserFactory.java]: Is it possible to move to /v2 and rename ParserFactory ?

File: src/test/java/org/symphonyoss/integration/webhook/jira/parser/v2/IssueCreatedMetadataParserTest.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+  private static final String FILE_EXPECTED_ISSUE_CREATED_TEMPLATE_MESSAGEML =
+      "templates/templateIssueCreated.xml";
+
+  @Mock
+  private UserService userService;
+
+  private MetadataParser parser;
+
+  private String expectedTemplateFile = "<messageML>\n"
```
Comment (crepache) [inline src/test/java/org/symphonyoss/integration/webhook/jira/parser/v2/IssueCreatedMetadataParserTest.java]: Is it possible move for one File ?

File: src/test/java/org/symphonyoss/integration/webhook/jira/parser/V1ParserFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Unit test for {@link V1ParserFactory}
+ * Created by rsanchez on 23/03/17.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class V1ParserFactoryTest {
```
Comment (crepache) [inline src/test/java/org/symphonyoss/integration/webhook/jira/parser/V1ParserFactoryTest.java]: Is it possible to move to /v1 and rename ParserFactoryTest ?

**Comment by ecarrenho-daitan** [comment]
@dnsymphony please review

**Comment by dnsymphony** [comment]
In general, lacking useful javadocs.  There are a lot of classes working together (parsers, factories, adapters, etc.) and not enough javadoc explaining how they work.  I can't review this since I don't understand the intent.

**Review by crepache — APPROVED** [review]

**Comment by robson-sanchez** [comment]
Improved the documentation and also removed part of the code to be reused in the other parsers reducing the number of artifacts to be reviewed.

@dnsymphony please check again

Complementary PR's: #28 and https://github.com/symphonyoss/App-Integrations-Commons/pull/44

File: src/main/java/org/symphonyoss/integration/webhook/jira/parser/v2/V2ParserFactory.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    return new ArrayList<JiraParser>(beans);
+  }
+
+  @Override
+  public JiraParser getParser(JsonNode node) {
+    JiraParser result = super.getParser(node);
+
+    if (result == null) {
+      // Fallback use V1 Factory
+      return fallbackFactory.getParser(node);
```
Comment (dnsymphony) [inline src/main/java/org/symphonyoss/integration/webhook/jira/parser/v2/V2ParserFactory.java:63]: Does this work?  Won't this end up sending message ML V1 to the Agent at an endpoint that expects V2?

Comment (robson-sanchez) [reply src/main/java/org/symphonyoss/integration/webhook/jira/parser/v2/V2ParserFactory.java:63]: Yes, it does. The message object has the messageML version as an attribute and the component that post messages will check the messageML version to use the proper API version.

It was implemented to support the partial rollout of the events.

**Comment by robson-sanchez** [comment]
@dnsymphony please check again

**Review by dnsymphony — APPROVED** [review]

**Pull request merged**